### ISSUE TITLE
torch.compile "run at compile time" decorator.

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#133944]
### 🚀 The feature, motivation and pitch

Let's say I have something like this
```
@cached
def create_pi_expensive():
    return torch.ones((), device='cuda', dtype=torch.float16) * 3.14159265358979323846

def f(x):
    y = create_pi()
    return x + y
```
where we want to cache something at compile time and then reuse it.

Today, this is not so easy to do. If you use `@lru_cache` we completely ignore the decorator and run the function every time (not sure this is the right behavior fwiw). OTOH, even if you implement a manual caching mechanism, we will at the very least need to compile your function twice.

I think it would make sense to allow for a "comptime" annotation.

cc: @ezyang @yanboliang @anijain2305

**Comment by Chillee** [comment]
Discussing with @anijain2305, one idea we came up with is instead to have a `torch.compiler.bailout()` function - this would allow you to bail out on runs that aren't using compile.

**Comment by anijain2305** [comment]
So we have an API already which does something familiar - assume_constant_result

~~~
import torch


@torch._dynamo.assume_constant_result
def create_pi():
    return torch.ones((), device='cuda', dtype=torch.float16) * 3.14159265358979323846

def f(x):
    y = create_pi()
    return x + y

opt_fn = torch.compile(f)
opt_fn(torch.ones(4, device="cuda"))
~~~

**Comment by jansel** [comment]
Yes, this already exists.  We could double check it works as expected, but if it doesn't it is a bug.

**Comment by StrongerXi** [comment]
Yeah I think `assume_constant_result` does the job here.

In fact, `allow_in_graph` (or its enhanced version `nonstrict_trace`) should also work, although that's probably a (?)unintended side effect of aot-dispatcher..? cc @zou3519 

Should we close this then?

**Comment by zou3519** [comment]
allow_in_graph/nonstrict_trace don't do this for tensor operations (they would put the tensor operations into the graph instead of evaluate them at compile time). So this is something that we'd need assume_constant_result.

My Q for assume_constant_result is:
1) is this use case relatively common
2) if so, then we should should expose assume_constant_result (in the torch.compiler namespace) and consider naming it something like "evaluate_at_compile_time" (it took me some mental gymnastics to realize that's what assume_constant_result does)

**Comment by jansel** [comment]
If we keep it, I'd support renaming it.
